infer/types/asset.go

```diff
@@ -0,0 +1,22 @@
+// Copyright 2022, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package types
+
+import "github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+
+type AssetOrArchive struct {
+	Asset   *resource.Asset   `pulumi:"asset,optional"`
+	Archive *resource.Archive `pulumi:"archive,optional"`
```

Comment:
I thought we discussed using a special sig here to ensure we found `AssetOrArchive` and not some other type.
```suggestion
	Asset   *resource.Asset   `pulumi:"somereallylongsigforassets,optional"`
	Archive *resource.Archive `pulumi:"somereallylongsigforarchives,optional"`
```